framework/source/class/qx/ui/table/cellrenderer/AbstractImage.js

```diff
@@ -251,6 +264,80 @@ qx.Class.define("qx.ui.table.cellrenderer.AbstractImage",
     },
 
 
+    /**
+     * Detects whether there is a high-resolution image available.
+     * A high-resolution image is assumed to have the same file name as
+     * the parameter source, but with a pixelRatio identifier before the file
+     * extension, like "@2x".
+     * Medium Resolution: "example.png", high-resolution: "example@2x.png"
+     *
+     * @param lowResImgSrc {String} source of the low resolution image.
+     * @return {String|Boolean} If a high-resolution image source.
+     */
+    _findHighResolutionSource: function(lowResImgSrc) {
+        if (lowResImgSrc === null)
+        {
+          return null;
+        }
+
+        var pixelRatioCandidates = ["3", "2", "1.5"];
+
+        // Calculate the optimal ratio, based on the rem scale factor of the application and the device pixel ratio.
+        var factor = parseFloat(qx.bom.client.Device.getDevicePixelRatio().toFixed(2));
+        if (factor <= 1) {
+            return false;
+        }
+
+        var i = pixelRatioCandidates.length;
+        while (i > 0 && factor > pixelRatioCandidates[--i]) {}
+
+        var hiResImgSrc;
+        var k;
+
+        // Search for best img with a higher resolution.
+        for (k = i; k >= 0; k--) {
+            hiResImgSrc = this._getHighResolutionSource(lowResImgSrc, pixelRatioCandidates[k]);
+            if (hiResImgSrc) {
+                return hiResImgSrc;
+            }
+        }
+
+        // Search for best img with a lower resolution.
+        for (k = i + 1; k < pixelRatioCandidates.length; k++) {
+            hiResImgSrc = this._getHighResolutionSource(lowResImgSrc, pixelRatioCandidates[k]);
+            if (hiResImgSrc) {
+                return hiResImgSrc;
+            }
+        }
+
+        return null;
+    },
+
+    /**
+     * Returns the source name for the high-resolution image based on the passed
+     * parameters.
+     * @param source {String} the source of the medium resolution image.
+     * @param pixelRatio {Number} the pixel ratio of the high-resolution image.
+     * @return {String} the high-resolution source name or null if no source could be found.
+     */
+    _getHighResolutionSource : function(source, pixelRatio) {
+        if (source === null)
+        {
+            return null;
+        }
+        var fileExtIndex = source.lastIndexOf('.');
+        if (fileExtIndex > -1) {
+            var pixelRatioIdentifier = "@" + pixelRatio + "x";
+            var candidate = source.slice(0, fileExtIndex) + pixelRatioIdentifier + source.slice(fileExtIndex);
+
+            if(qx.util.ResourceManager.getInstance().has(candidate)) {
+                return candidate;
+            }
+        }
+        return null;
+    },
+
+
```

Comment:
I think we should move `_findHighResolutionSource` and `_getHighResolutionSource` into `qx.util.ResourceManager` as public methods and use them here and in `qx.ui.basic.Image`.
The tests in `qx.test.ui.basic.Images.testHighResImage` should reveal any problems.

Reply:
+1